stats/stats.go

```diff
@@ -0,0 +1,85 @@
+package stats
+
+import "time"
+
+// RPCStats contains stats information about RPCs.
+type RPCStats interface {
+	isRPCStats()
+	// IsClient returns true if this RPCStats is from client side.
+	IsClient() bool
+}
+
+// Begin contains stats when an RPC begins.
+// FailFast is only valid if this Begin is from client side.
+type Begin struct {
+	// Client is true if this Begin is from client side.
+	Client bool
+	// BeginTime is the time when the RPC begins.
+	BeginTime time.Time
+}
+
+// IsClient indicates if the stats information is from client side.
+func (s *Begin) IsClient() bool { return s.Client }
+
+func (s *Begin) isRPCStats() {}
+
+// InPayload contains the information for an incoming payload.
+type InPayload struct {
+	// Client is true if this InPayload is from client side.
+	Client bool
+	// Payload is the payload with original type.
+	Payload interface{}
+	// Data is the serialized message payload.
+	Data []byte
+	// Length is the length of uncompressed data.
+	Length int
+	// WireLength is the length of data on wire (compressed, signed, encrypted).
+	WireLength int
+	// RecvTime is the time when the payload is received.
+	RecvTime time.Time
+}
+
+// IsClient indicates if the stats information is from client side.
+func (s *InPayload) IsClient() bool { return s.Client }
+
+func (s *InPayload) isRPCStats() {}
+
+// OutPayload contains the information for an outgoing payload.
+type OutPayload struct {
```

Comment:
Should InPayload and OutPayload be the same type (considering they carry pretty much the same information)?

Reply:
No as a type switch is used to determine the type of event that has just occurred in the plugin.

Reply:
then, embed a common type in both?

```
type OutPayload struct {
  payload
}
```
etc

Reply:
I would have to make the `payload` type an exported type, which it shouldn't be, because nothing should ever use the `payload` type itself and only the `InPayload`/`OutPayload` types should be used. And making the `payload` unexported means that I would need to provide constructor methods for the In/OutPayload types so access the `payload` type that is embedded in them.

Reply:
It should work with an unexported type with exported fields. There is an example here:

https://www.ardanlabs.com/blog/2014/03/exportedunexported-identifiers-in-go.html

Reply:
ok, I've thought of it and it's good like this (it's not possible to document unexported type exported fields properly).